src/gpstools/doc/gpst.1
.\"     Title: gpst
.\"    Author: [see the "Author" section]
.\" Generator: DocBook XSL Stylesheets v1.74.0 <http://docbook.sf.net/>
.\"      Date: $Id$
.\"    Manual: [FIXME: manual]
.\"    Source: [FIXME: source]
.\"  Language: English
.\"
.TH "GPST" "1" "$Id$" "[FIXME: source]" "[FIXME: manual]"
.\" -----------------------------------------------------------------
.\" * (re)Define some macros
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" toupper - uppercase a string (locale-aware)
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de toupper
.tr aAbBcCdDeEfFgGhHiIjJkKlLmMnNoOpPqQrRsStTuUvVwWxXyYzZ
\\$*
.tr aabbccddeeffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH-xref - format a cross-reference to an SH section
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de SH-xref
.ie n \{\
.\}
.toupper \\$*
.el \{\
\\$*
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH - level-one heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SH
.\" put an extra blank line of space above the head in non-TTY output
.if t \{\
.sp 1
.\}
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[an-margin]u
.ti 0
.HTML-TAG ".NH \\n[an-level]"
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
\." make the size of the head bigger
.ps +3
.ft B
.ne (2v + 1u)
.ie n \{\
.\" if n (TTY output), use uppercase
.toupper \\$*
.\}
.el \{\
.nr an-break-flag 0
.\" if not n (not TTY), use normal case (not uppercase)
\\$1
.in \\n[an-margin]u
.ti 0
.\" if not n (not TTY), put a border/line under subheading
.sp -.6
\l'\n(.lu'
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SS - level-two heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SS
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[IN]u
.ti \\n[SN]u
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.ps \\n[PS-SS]u
\." make the size of the head bigger
.ps +2
.ft B
.ne (2v + 1u)
.if \\n[.$] \&\\$*
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BB/BE - put background/screen (filled box) around block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BB
.if t \{\
.sp -.5
.br
.in +2n
.ll -2n
.gcolor red
.di BX
.\}
..
.de EB
.if t \{\
.if "\\$2"adjust-for-leading-newline" \{\
.sp -1
.\}
.br
.di
.in
.ll
.gcolor
.nr BW \\n(.lu-\\n(.i
.nr BH \\n(dn+.5v
.ne \\n(BHu+.5v
.ie "\\$2"adjust-for-leading-newline" \{\
\M[\\$1]\h'1n'\v'+.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.el \{\
\M[\\$1]\h'1n'\v'-.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.in 0
.sp -.5v
.nf
.BX
.in
.sp .5v
.fi
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BM/EM - put colored marker in margin next to block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BM
.if t \{\
.br
.ll -2n
.gcolor red
.di BX
.\}
..
.de EM
.if t \{\
.br
.di
.ll
.gcolor
.nr BH \\n(dn
.ne \\n(BHu
\M[\\$1]\D'P -.75n 0 0 \\n(BHu -(\\n[.i]u - \\n(INu - .75n) 0 0 -\\n(BHu'\M[]
.in 0
.nf
.BX
.in
.fi
.\}
..
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "Name"
gpst \- Converts between various GPS formats\&.
.SH "Synopsis"
.fam C
.HP \w'\fBgpst\fR\ 'u
\fBgpst\fR [\fIoptions\fR] [\fIfile\fR...]
.fam
.fam C
.HP \w'\fBgpst\fR\ 'u
\fBgpst\fR {\-S} [\fIoptions\fR] [\fIfile\fR...]
.fam
.fam C
.HP \w'\fBgpst\fR\ 'u
\fBgpst\fR {\-u} [\fIoptions\fR] [\fIfile\fR...]
.fam
.SH "Description"
.PP
\fB\-\-chronology\fR
.RS 4
Check for broken chronology, warn about entries with an old timestamp\&.
.RE
.PP
\fB\-d\fR, \fB\-\-skip\-dups\fR
.RS 4
If succeeding identical coordinates are found, print only the first one and ignore the rest\&.
.RE
.PP
\fB\-e\fR, \fB\-\-epoch\fR
.RS 4
Use seconds since 1970\-01\-01 00:00:00 GMT as date format\&.
.RE
.PP
\fB\-\-fix\fR
.RS 4
Comment out entries which is obviously wrong\&. Use together with
\fB\-\-chronology\fR
to fix those kind of errors\&. Does not work with GPX output yet\&.
.RE
.PP
\fB\-\-from\-date \fR\fB\fIx\fR\fR
.RS 4
Used by the
\FCpgwupd\F[]
format\&. Specifies from which date waypoints should be updated\&. No checks for valid date format here, let
PostgreSQL
take care of that\&. All variants it understands can be used here\&.
.RE
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Send a brief help summary to stdout\&.
.RE
.PP
\fB\-\-inside\fR
.RS 4
Print only trackpoints inside a rectangle specified by
\fB\-\-pos1\fR
and
\fB\-\-pos2\fR\&.
.RE
.PP
\fB\-n\fR, \fB\-\-undefined \fR\fB\fIx\fR\fR
.RS 4
Use
\fIx\fR
as undefined value\&.
.RE
.PP
\fB\-\-near\fR
.RS 4
Add names of the three closest waypoints to the trackpoint\&. Unfinished and experimental, needs
\fBgpsbabel\fR(1), which is called from the program as
\fB/usr/local/bin/gpsbabel\fR\&.
.RE
.PP
\fB\-o\fR, \fB\-\-output\-format \fR\fB\fIx\fR\fR
.RS 4
Use output format
\fIx\fR:
.PP
\fBclean\fR
.RS 4
TAB\-separated format with longitude, latitude and altitude\&.
.RE
.PP
\fBcsv\fR
.RS 4
TAB\-separated format:
.PP
date
.RS 4
Date of trackpoint\&. Format:
\FC\fIyyyy\fR\-\fImm\fR\-\fIdd\fRT\fIhh\fR:\fImm\fR:\fIss\fRZ\F[]
unless the
\fB\-\-epoch\fR
option is specified\&. If so, it\(cqs represented as number of seconds since 1970\-01\-01 00:00:00 UTC\&.
.RE
.PP
longitude
.RS 4
Longitude, decimal degrees\&.
.RE
.PP
latitude
.RS 4
Latitude, decimal degrees\&.
.RE
.PP
altitude
.RS 4
Altitude in metres above sea level\&. Optional\&. If not present, the corresponding TAB is also printed, to keep the number of columns intact\&.
.RE
.RE
.PP
\fBgpsml\fR (Default)
.RS 4
Line\-based XML which makes it easy to edit and grep\&. In the past, there was this big plan about making this the default storage format, so it was declared as The Honourable Default Output Format\&. But using GPX is the way to go, so this will probably change in the future\&. It\(cqs way too
de facto
and widely used, so storing GPS data in anything else is sheer waste of work\&. Probably not finished yet\&.
.RE
.PP
\fBgpstrans\fR
.RS 4
The format generated by
\fBgpstrans\fR(1)\&.
.RE
.PP
\fBgpx\fR
.RS 4
Not complete yet, but good enough in most cases\&.
.RE
.PP
\fBpgtab\fR
.RS 4
TAB\-separated format for import into
PostgreSQL\&. Undefined values are written as
\FC\eN\F[]
if the
\fB\-\-undefined\fR
option is not specified\&.
.PP
date
.RS 4
Format:
\FC\fIyyyy\fR\-\fImm\fR\-\fIdd\fRT\fIhh\fR:\fImm\fR:\fIss\fRZ\F[]
.RE
.PP
coor
.RS 4
Format:
\FC(\fIlat\fR,\fIlon\fR)\F[]
.RE
.PP
ele
.RS 4
Altitude in metres above sea level\&.
.RE
.PP
sted
.RS 4
Short name of closest waypoint\&. Calculated in
PostgreSQL, so this value is always
\FC\eN\F[]\&.
.RE
.PP
dist
.RS 4
Distance in decimal degrees to closest waypoint\&. Calculated in
PostgreSQL, so this value is always
\FC\eN\F[]\&.
.RE
.PP
description
.RS 4
Description of waypoint\&. Calculated in
PostgreSQL, so this value is always
\FC\eN\F[]\&.
.RE
.PP
avst
.RS 4
Distance to trackpoint in decimal degrees from a defined
\(lqhome\(rq
position\&. This is calculated in
PostgreSQL, so the field value is always
\FC\eN\F[]\&.
.RE
.RE
.PP
\fBpgwtab\fR
.RS 4
Output waypoints on a TAB\-separated format:
.PP
coor
.RS 4
Format:
\FC(\fIlat\fR,\fIlon\fR)\F[]
.RE
.PP
name
.RS 4
Title of waypoint\&.
.RE
.PP
ele
.RS 4
Elevation in metres above sea level\&.
.RE
.PP
type
.RS 4
Type of waypoint, same as the GPX
<type>
element\&.
.RE
.PP
time
.RS 4
Timestamp of waypoint\&.
.RE
.PP
cmt
.RS 4
Short waypoint description, used by the GPS as a comment\&. Same as the GPX
<cmt>
element\&.
.RE
.PP
descr
.RS 4
Long description, not used by the GPS\&. Same as the GPX
<desc>
element\&.
.RE
.PP
src
.RS 4
Source of data, same as the
<src>
GPX element\&.
.RE
.PP
sym
.RS 4
GPS symbol name\&. Same as the
<sym>
GPX element\&.
.RE
.RE
.PP
\fBpgwupd\fR
.RS 4
Output waypoints as SQL transactions which updates the
sted
and
dist
columns\&.
.RE
.PP
\fBposcount\fR
.RS 4
Creates a 3D plot where areas with many trackpoints are higher than areas with less track points\&.
.RE
.PP
\fBps\fR
.RS 4
Postscript\&. Unfinished\&.
.RE
.PP
\fBsvg\fR
.RS 4
Unfinished\&.
.RE
.PP
\fBxgraph\fR
.RS 4
Input format for
\fBxgraph\fR(1)\&.
.RE
.PP
\fBygraph\fR
.RS 4
Input format for
\fBygraph\fR(1)\&.
.RE
.RE
.PP
\fB\-\-outside\fR
.RS 4
Print only trackpoints outside a rectangle specified by
\fB\-\-pos1\fR
and
\fB\-\-pos2\fR\&.
.RE
.PP
\fB\-\-pos1 \fR\fB\fIx1\fR\fR\fB,\fR\fB\fIy1\fR\fR, \fB\-\-pos2 \fR\fB\fIx2\fR\fR\fB,\fR\fB\fIy2\fR\fR
.RS 4
Specifies corners of an area rectangle used by the
\fB\-\-inside\fR
and
\fB\-\-outside\fR
options\&. The
\fIx\fR
and
\fIy\fR
values are specified as decimal degrees, negative for west or south\&.
.RE
.PP
\fB\-r\fR, \fB\-\-require \fR\fB\fIx\fR\fR
.RS 4
Specify requirements for trackpoints to be written\&.
\fIx\fR
is a string with the following flags:
.PP
\fBe\fR
.RS 4
Print only waypoints which have an elevation\&.
.RE
.PP
\fBp\fR
.RS 4
Print only waypoints which have a position\&.
.RE
.PP
\fBt\fR
.RS 4
Print only waypoints which have a timestamp\&.
.RE
.RE
.PP
\fB\-R\fR, \fB\-\-round \fR\fB\fIx\fR\fR\fB=\fR\fB\fIy\fR\fR\fB[,\fIx2\fR=\fIy2\fR[\&.\&.\&.]]\fR
.RS 4
Round trackpoint element
\fIx\fR
to
\fIy\fR
decimals\&.
.sp
Example:
\fB\-\-round lat=4,lon=5,ele=1\fR
.RE
.PP
\fB\-s\fR, \fB\-\-short\-date\fR
.RS 4
Use short date format\&.
.RE
.PP
\fB\-S\fR, \fB\-\-save\-to\-file \fR\fB\fIx\fR\fR
.RS 4
Save the unconverted data to a file with a filename starting with the timestamp of the first trackpoint\&. The parameter string
\fIx\fR
is added at the end of the filename\&. For the time being this option will ignore all other options\&.
.if n \{\
.sp
.\}
.RS 4
.BM yellow
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
If several files are specified on the command line, all data will be saved into only one file\&. This behaviour may change in the future\&.
.sp .5v
.EM yellow
.RE
.RE
.PP
\fB\-t\fR, \fB\-\-create\-breaks\fR
.RS 4
Create breaks in track between points with a difference more than the number of seconds specified by the
\fBPAUSE_LIMIT\fR
variable\&.
.RE
.PP
\fB\-u\fR, \fB\-\-comment\-out\-dups\fR
.RS 4
Comment out following data with identical position values, only print first entry\&.
.RE
.PP
\fB\-v\fR, \fB\-\-verbose\fR
.RS 4
Verbose output, can be repeated\&.
.RE
.PP
\fB\-\-version\fR
.RS 4
Print program version information\&.
.RE
.PP
\fB\-w\fR, \fB\-\-strip\-whitespace\fR
.RS 4
Strip all unnecessary whitespace\&.
.RE
.PP
\fB\-y\fR, \fB\-\-double\-y\-scale\fR
.RS 4
Double Y scale (latitude)\&.
.RE
.PP
\fB\-\-debug\fR
.RS 4
Print debugging messages\&.
.RE
.SH "Bugs and things that need to be done"
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Pretty incomplete in some areas\&. Some of the source formats are undocumented and thus incomplete\&. Some functionality is not working properly, for example the Postscript output\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Remove hardcodings, for example the
\(lqhome\(rq
position used to calculate the
avst
database column\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Translate database column names to English\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Add more tests to cover all command\-line options\&.
.RE
.SH "Author"
.PP
Made by
Øyvind A\&. Holm
\FCsunny@sunbase\&.org\F[]\&.
.SH "Copyright"
.PP
Copyleft \(co
Øyvind A\&. Holm
\FCsunny@sunbase\&.org\F[]\&. This is free software; see the file
\FCCOPYING\F[]
for legalese stuff\&.
.SH "License"
.PP
This program is free software; you can redistribute it and/or modify it under the terms of the
GNU General Public License
as published by the
Free Software Foundation; either version 2 of the License, or (at your option) any later version\&.
.PP
This program is distributed in the hope that it will be useful, but
\fIWITHOUT ANY WARRANTY\fR; without even the implied warranty of
\fIMERCHANTABILITY\fR
or
\fIFITNESS FOR A PARTICULAR PURPOSE\fR\&. See the GNU General Public License for more details\&.
.PP
You should have received a copy of the GNU General Public License along with this program; if not, write to
.sp
.if n \{\
.RS 4
.\}
.fam C
.ps -1
.nf
The Free Software Foundation, Inc\&.
59 Temple Place, Suite 330
Boston, MA 02111\-1307
USA
.fi
.fam
.ps +1
.if n \{\
.RE
.\}
.SH "See also"
.PP
\fBgpsbabel\fR(1)


branches/gpst.gpxfix/GPST.pm
package GPST;

#=======================================================================
# $Id$
#
# Character set: UTF-8
# ©opyleft 2002– Øyvind A. Holm <sunny@sunbase.org>
# License: GNU General Public License, see end of file for legal stuff.
#=======================================================================

use strict;
use warnings;

use GPSTdebug;
use GPSTgeo;

BEGIN {
    use Exporter ();
    our ($VERSION, @ISA, @EXPORT, @EXPORT_OK, %EXPORT_TAGS);

    my $rcs_id = '$Id$';
    push(@main::version_array, $rcs_id);
    $VERSION = ($rcs_id =~ / (\d+) /, $1);

    @ISA = qw(Exporter);
    @EXPORT = qw(&trackpoint &postgresql_copy_safe @wpt_elems);
    %EXPORT_TAGS = ();
}
our @EXPORT_OK;

our @wpt_elems = (
    # Position info
    "ele", "time", "magvar", "geoidheight",
    # Description info
    "name", "cmt", "desc", "src", "link", "sym", "type",
    # Accuracy info
    "fix", "sat", "hdop", "vdop", "pdop",
    "ageofdgpsdata", "dgpsid",
    # Extensions
    "extensions"
);

our $Spc = " ";

sub trackpoint {
    # Receive a hash and return a trackpoint as a string {{{
    my %Dat = @_;

    defined($Dat{'what'}) || return(undef);
    defined($Dat{'format'}) || return(undef);
    defined($Dat{'error'}) || return(undef);

    defined($Dat{'year'}) || ($Dat{'year'} = 0);
    defined($Dat{'month'}) || ($Dat{'month'} = 0);
    defined($Dat{'day'}) || ($Dat{'day'} = 0);
    defined($Dat{'hour'}) || ($Dat{'hour'} = "");
    defined($Dat{'min'}) || ($Dat{'min'} = "");
    defined($Dat{'sec'}) || ($Dat{'sec'} = "");
    my $print_time = (
        !$Dat{'year'} ||
        !$Dat{'month'} ||
        !$Dat{'day'} ||
        !length($Dat{'hour'}) ||
        !length($Dat{'min'}) ||
        !length($Dat{'sec'})
    ) ? 0 : 1;

    if (
        ("$Dat{'year'}$Dat{'month'}$Dat{'day'}$Dat{'hour'}$Dat{'min'}" =~
        /[^\d]/) || ($Dat{'sec'} =~ /[^\d\.]/)
    ) {
        ($print_time = 0);
    }
    "$Dat{'lat'}$Dat{'lon'}" =~ /[^\d\.\-\+]/ && return(undef);

    defined($Dat{'lat'}) || ($Dat{'lat'} = "");
    defined($Dat{'lon'}) || ($Dat{'lon'} = "");
    defined($Dat{'ele'}) || ($Dat{'ele'} = "");
    defined($Dat{'desc'}) || ($Dat{'desc'} = "");
    defined($Dat{'extensions'}) || ($Dat{'extensions'} = "");

    my $Retval = "";

    if ($Dat{'what'} eq "tp") {
        my $err_str = length($Dat{'error'}) ? $Dat{'error'} : "";
        if ($Dat{'format'} eq "gpsml") {
            # {{{
            my $Elem = length($err_str) ? "etp" : "tp";
            $Retval .= join("",
                    $print_time
                        ? sprintf("<time>%04u-%02u-%02uT" .
                                  "%02u:%02u:%02gZ</time> ",
                                   $Dat{'year'}, $Dat{'month'}, $Dat{'day'},
                                   $Dat{'hour'}, $Dat{'min'}, $Dat{'sec'}*1.0
                          )
                        : "",
                    (length($Dat{'lat'}))
                        ? "<lat>" . $Dat{'lat'}*1.0 . "</lat> "
                        : "",
                    (length($Dat{'lon'}))
                        ? "<lon>" . $Dat{'lon'}*1.0 . "</lon> "
                        : "",
                    (length($Dat{'ele'}))
                        ? "<ele>" . $Dat{'ele'}*1.0 . "</ele> "
                        : "",
                    (length($Dat{'desc'}))
                        ? sprintf("<desc>%s</desc> ",
                                  $Dat{'desc'})
                        : ""
            );
            length($Retval) &&
                ($Retval = sprintf("<%s%s> %s</%s>\n",
                                   $Elem,
                                   length($err_str) ? " err=\"$err_str\"" : "",
                                   $Retval,
                                   $Elem)
            );
            # }}}
        } elsif($Dat{'format'} eq "gpx") {
            # {{{
            my $lat_str = length($Dat{'lat'}) ? " lat=\"$Dat{'lat'}\"" : "";
            my $lon_str = length($Dat{'lon'}) ? " lon=\"$Dat{'lon'}\"" : "";
            my ($estr_begin, $estr_end) =
               (         "",        "");
            if (length($err_str)) {
                $estr_begin = "<!-- ";
                $Dat{'extensions'} .= "<error>$err_str</error>";
                $estr_end = " -->";
            }
            if (length("$lat_str$lon_str$Dat{'ele'}")) {
                $Retval .=
                join("",
                    "$Spc$Spc$Spc$Spc$Spc$Spc",
                    $estr_begin,
                    "<trkpt$lat_str$lon_str>",
                    "$Spc",
                    length($Dat{'ele'})
                        ? "<ele>$Dat{'ele'}</ele>$Spc"
                        : "",
                    $print_time
                        ? "<time>" .
                          "$Dat{'year'}-$Dat{'month'}-$Dat{'day'}T" .
                          "$Dat{'hour'}:$Dat{'min'}:$Dat{'sec'}Z" .
                          "</time>$Spc"
                        : "",
                    length($Dat{'extensions'})
                        ? "<extensions>" .
                          $Spc .
                          $Dat{'extensions'} .
                          $Spc .
                          "</extensions>$Spc"
                        : "",
                    "</trkpt>$estr_end\n"
                );
            }
            # }}}
        } elsif($Dat{'format'} eq "clean") {
            $Retval .= "$Dat{'lon'}\t$Dat{'lat'}\t$Dat{'ele'}\n";
        } elsif($Dat{'format'} eq "xgraph") {
            if (length($Dat{'lat'}) && length($Dat{'lon'})) {
                $Retval .= "$Dat{'lon'} $Dat{'lat'}\n";
            }
        } elsif ($Dat{'format'} eq "pgtab") {
            $Retval .= join("\t",
                $Dat{'year'}
                    ? "$Dat{'year'}-$Dat{'month'}-$Dat{'day'}T" .
                      "$Dat{'hour'}:$Dat{'min'}:$Dat{'sec'}Z"
                    : '\N', # date
                (length($Dat{'lat'}) && length($Dat{'lon'}))
                    ? "($Dat{'lat'},$Dat{'lon'})"
                    : '\N', # coor
                length($Dat{'ele'}) ? $Dat{'ele'} : '\N', # ele
                '\N', # sted
                '\N', # dist
                '\N', # description
                '\N' # avst
            ) . "\n";
        } elsif ($Dat{'format'} eq "gpstrans") {
            # {{{
            my ($gpt_lat, $gpt_lon) =
               (ddd_to_dms($Dat{'lat'}), ddd_to_dms($Dat{'lon'}));
            if ($print_time) {
                $Retval .= "T\t$Dat{'month'}/$Dat{'day'}/$Dat{'year'} " .
                         "$Dat{'hour'}:$Dat{'min'}:$Dat{'sec'}\t" .
                         "$gpt_lat\t$gpt_lon\n";
            } else {
                $Retval .= "T\t00/00/00 00:00:00\t$gpt_lat\t$gpt_lon\n";
            }
            # }}}
        } else {
            $Retval = undef;
        }
    } else {
        $Retval = undef;
    }
    return $Retval;
    # }}}
}

sub postgresql_copy_safe {
    # {{{
    my $Str = shift;
    $Str =~ s/\\/\\\\/gs;
    $Str =~ s/\n/\\n/gs;
    $Str =~ s/\r/\\r/gs;
    $Str =~ s/\t/\\t/gs;
    return($Str);
    # }}}
}

1;
